### DISCUSSION TITLE
Default TemplateSlots configurable in custom Layouts

### DISCUSSION THREAD

**Discussion by jhholm** [microsoft-search/pnp-modern-search#2411] [Ideas]
I've been working with SharePoint OTB Display Templates since SP 2013 and one feature that I've missed in PnP Modern Search is the capability to set the default template slots for a custom layout. Currently I am just documenting the available slots and preferred default values in the custom layout property pane fields configuration. Thus this is a bit tedious to configure.

I'm happy to create a pull request for this if this seems like a good idea develop further.

My initial ideas:
* Add an optional **defaultTemplateSlots: ITemplateSlot[]** property to **ILayoutDefinition** in **@pnp/modern-search-extensibility**
* Add a button to get template slots from layout (see below - i just used the reset button for mockup)
  * Could be grayed out, if the property is empty
* Add a similar reset button to set template defaults (see below)

I think the UI and UX are really important how to handle this intuitively. 

![image](https://user-images.githubusercontent.com/13876124/193838511-381e938c-5327-4aff-9dea-44fd841c412a.png)
Add setting template slots from layout

![image](https://user-images.githubusercontent.com/13876124/193839185-6a44be06-89ee-42b1-a651-6762553248b8.png)
Add reset button here to set dataSource defaults

**Comment by FranckyC** [comment]
Hi @jhholm, you're right, this is not possible right now with the current implementation. The slots are specific to a data source, not a layout. By definition, a layout is not tied to a specific data source so having default slot values is not really an option.

However, having a way to define slots exposed by a specific layout and highlight them in the slots configuration could help user to set up the right fields.

I'll tag it as an enhancement,

Thanks!

**Reply by jhholm** [reply]
I guess we could change the semantics here. Instead of using **defaultSlots**, we could use **slotsUsed**. Then providing a button to append the list of slot names used without populating the slot fields to the currently configured slots. Does this sound plausible?

We could additionally define for the layout, that if you are using dataSource X, use slot fields A, and for dataSource Y, use slot fields B. Unknown dataSource would just get the fields empty. Or does this sound too complex?

**Reply by jhholm** [reply]
My use case here is providing a layout that looks like SharePoint OTB News Web Part. Currently custom layouts like this require quite a lot of manual configuration and it would be great to get as much as possible preconfigured declaratively.

Currently the workflow goes like this:
1. Configure the Search Results Web Part to get the results wanted
2. Configure and set the custom layout
2. Add the additional managed properties to fetched for dataSource SharePoint (usually per customer)
3. Add the additional slots names needed (Views, PublishedDate, IsOrganizationalNewsSite etc)
4. Configure the slot fields from managed properties to the slot names
